src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs

```diff
@@ -0,0 +1,87 @@
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Localization;
+using OrchardCore.ContentLocalization.Services;
+using OrchardCore.DisplayManagement.ModelBinding;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Localization;
+using OrchardCore.Modules;
+
+namespace OrchardCore.ContentLocalization.Controllers
+{
+    [Feature("OrchardCore.ContentLocalization.ContentCulturePicker")]
+    public class ContentCulturePickerController : Controller, IUpdateModel
+    {
+        private readonly ILocalizationService _locationService;
+        private readonly IContentCulturePickerService _culturePickerService;
+        private readonly string _tenantPrefix;
+
+        public IHtmlLocalizer T { get; }
+
+        public ContentCulturePickerController(
+            IHtmlLocalizer<AdminController> localizer,
+            ILocalizationService locationService,
+            IContentCulturePickerService culturePickerService,
+            ShellSettings shellSettings
+           )
+        {
+            T = localizer;
+            _locationService = locationService;
+            _tenantPrefix = "/" + shellSettings.RequestUrlPrefix;
+            _culturePickerService = culturePickerService;
+        }
+
+        [HttpGet]
+        public async Task<IActionResult> RedirectToLocalizedContent(string targetCulture, PathString contentItemUrl)
+        {
+            // Invariant culture name is empty so a null value is bound.
+            targetCulture = targetCulture ?? "";
+
+            if (contentItemUrl == "/Error")
+            {
+                return NoContent();
+            }
+
+            if (!contentItemUrl.HasValue)
+            {
+                contentItemUrl = "/";
+            }
+
+            var supportedCultures = await _locationService.GetSupportedCulturesAsync();
+
+            if (!supportedCultures.Any(t => t == targetCulture))
+            {
+                return NoContent();
```

Comment:
Never return NoContent, the same page instead

Reply:
If so, do we need to preserve an eventual queryString ? Same question when we do a redirection. At some point we were passing the queryString to the controller, maybe not important when we do a redirection, but maybe yes when we stay on the same page.